libsql-replication/src/meta.rs

```diff
@@ -0,0 +1,161 @@
+use std::io::{ErrorKind, SeekFrom};
+use std::mem::size_of;
+use std::path::Path;
+use std::str::FromStr;
+
+use bytemuck::{bytes_of, try_pod_read_unaligned, Pod, Zeroable};
+use tokio::fs::File;
+use tokio::io::{AsyncRead, AsyncReadExt, AsyncSeekExt, AsyncWriteExt};
+use tokio::pin;
+use uuid::Uuid;
+
+use crate::frame::FrameNo;
+use crate::rpc::replication::HelloResponse;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Invalid meta file")]
+    InvalidMetaFile,
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Invalid log id")]
+    InvalidLogId,
+    #[error("Replica log incompatible with log from handshake")]
+    LogIncompatible,
+    #[error("Failed to commit replication index")]
+    FailedToCommit(std::io::Error),
+}
+
+#[repr(C)]
+#[derive(Debug, Pod, Zeroable, Clone, Copy)]
+pub struct WalIndexMetaData {
+    /// id of the replicated log
+    log_id: u128,
+    /// committed frame index
+    pub committed_frame_no: FrameNo,
+    _padding: u64,
+}
+
+impl WalIndexMetaData {
+    async fn read(file: impl AsyncRead) -> Result<Option<Self>, Error> {
+        pin!(file);
+        let mut buf = [0; size_of::<WalIndexMetaData>()];
+        let meta = match file.read_exact(&mut buf).await {
+            Ok(_) => {
+                let meta: Self =
+                    try_pod_read_unaligned(&buf).map_err(|_| Error::InvalidMetaFile)?;
+                Some(meta)
+            }
+            Err(e) if e.kind() == ErrorKind::UnexpectedEof => None,
+            Err(e) => Err(e)?,
+        };
+
+        Ok(meta)
+    }
+}
+
+pub struct WalIndexMeta {
+    file: File,
+    data: Option<WalIndexMetaData>,
+}
+
+impl WalIndexMeta {
+    pub async fn open(db_path: &Path) -> Result<Self, Error> {
+        let path = db_path.join("client_wal_index");
+
+        tokio::fs::create_dir_all(db_path).await?;
+
+        let mut file = tokio::fs::OpenOptions::new()
+            .create(true)
+            .read(true)
+            .write(true)
+            .open(&path)
+            .await?;
+
+        let data = WalIndexMetaData::read(&mut file).await?;
+
+        Ok(Self { file, data })
+    }
+
+    /// attempts to merge two meta files.
```

Comment:
I find this comment and function name very confusing. Maybe it dives too deeply into implementation details? I don't see any two files taken as parameters, just a hello response, so maybe this is actually more a `init_wal_index_metadata_file` or something? The implementation suggests it simply either initializes a wal metadata file or tries to reuse an existing one.

Reply:
yeah, it's not great, improving the name a bit, but this is going away soon, this meta will be stored in page 1